File: virtual/GR-06/codigo/lu/lu_matrix_inversion.py
import numpy as np
import sys
import time
import os

def leer_csv(nombre_archivo):
    try:
        matriz = np.loadtxt(nombre_archivo, delimiter=",")
        if matriz.shape[0] != matriz.shape[1]:
            raise ValueError("La matriz no es cuadrada")
        return matriz
    except Exception as e:
        print(f"Error al leer el archivo CSV: {e}")
        return None

def lu_decomposition(A):
    N = A.shape[0]
    L = np.zeros((N, N))
    U = np.zeros((N, N))

    for i in range(N):
        for k in range(i, N):
            suma = sum(L[i][j] * U[j][k] for j in range(i))
            U[i][k] = A[i][k] - suma

        for k in range(i, N):
            if i == k:
                L[i][i] = 1.0
            else:
                suma = sum(L[k][j] * U[j][i] for j in range(i))
                L[k][i] = (A[k][i] - suma) / U[i][i]

    return L, U

def forward_substitution(L, b):
    N = L.shape[0]
    y = np.zeros(N)
    for i in range(N):
        y[i] = b[i] - np.dot(L[i, :i], y[:i])
    return y

def backward_substitution(U, y):
    N = U.shape[0]
    x = np.zeros(N)
    for i in reversed(range(N)):
        x[i] = (y[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i][i]
    return x

def invertir_matriz(A):
    N = A.shape[0]
    L, U = lu_decomposition(A)
    A_inv = np.zeros_like(A)

    for i in range(N):
        e = np.zeros(N)
        e[i] = 1.0
        y = forward_substitution(L, e)
        x = backward_substitution(U, y)
        A_inv[:, i] = x

    return A_inv

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Uso: python3 lu_matrix_inversion.py <dimensión>")
        sys.exit(1)

    try:
        N = int(sys.argv[1])
        if N <= 0:
            raise ValueError()
    except ValueError:
        print("Error: La dimensión debe ser un entero positivo")
        sys.exit(1)

    archivo = f"../../matrices/matriz_{N}x{N}_invertible.csv"

    A = leer_csv(archivo)
    if A is not None:
        start_time = time.time()
        A_inv = invertir_matriz(A)
        end_time = time.time()

        if N <= 10:
            print("Matriz inversa A^-1:")
            print(np.round(A_inv, 4))
        else:
            np.savetxt("inversa.csv", A_inv, delimiter=",", fmt="%.10f")
            print("Matriz inversa guardada en 'inversa.csv' (no se muestra por ser mayor de 10x10).")

        print(f"\nTiempo de ejecución: {end_time - start_time:.6f} segundos")

File: virtual/GR-06/codigo/lu/lu_matrix_inversion.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>

double** allocate_matrix(int n) {
    double** mat = malloc(n * sizeof(double*));
    for (int i = 0; i < n; i++)
        mat[i] = calloc(n, sizeof(double));
    return mat;
}

void free_matrix(double** mat, int n) {
    for (int i = 0; i < n; i++)
        free(mat[i]);
    free(mat);
}

void lu_decomposition(double** A, double** L, double** U, int N) {
    for (int i = 0; i < N; i++) {
        for (int k = i; k < N; k++) {
            double sum = 0.0;
            for (int j = 0; j < i; j++)
                sum += L[i][j] * U[j][k];
            U[i][k] = A[i][k] - sum;
        }
        for (int k = i; k < N; k++) {
            if (i == k)
                L[i][i] = 1.0;
            else {
                double sum = 0.0;
                for (int j = 0; j < i; j++)
                    sum += L[k][j] * U[j][i];
                L[k][i] = (A[k][i] - sum) / U[i][i];
            }
        }
    }
}

void forward_substitution(double** L, double* b, double* y, int N) {
    for (int i = 0; i < N; i++) {
        y[i] = b[i];
        for (int j = 0; j < i; j++)
            y[i] -= L[i][j] * y[j];
    }
}

void backward_substitution(double** U, double* y, double* x, int N) {
    for (int i = N - 1; i >= 0; i--) {
        x[i] = y[i];
        for (int j = i + 1; j < N; j++)
            x[i] -= U[i][j] * x[j];
        x[i] /= U[i][i];
    }
}

void invert_matrix(double** A, double** A_inv, int N) {
    double** L = allocate_matrix(N);
    double** U = allocate_matrix(N);
    lu_decomposition(A, L, U, N);
    for (int i = 0; i < N; i++) {
        double* e = calloc(N, sizeof(double));
        double* y = malloc(N * sizeof(double));
        double* x = malloc(N * sizeof(double));
        e[i] = 1.0;
        forward_substitution(L, e, y, N);
        backward_substitution(U, y, x, N);
        for (int j = 0; j < N; j++)
            A_inv[j][i] = x[j];
        free(e);
        free(y);
        free(x);
    }
    free_matrix(L, N);
    free_matrix(U, N);
}

void print_matrix(double** M, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++)
            printf("%8.4f ", M[i][j]);
        printf("\n");
    }
}

int contar_columnas(const char* linea) {
    int count = 1;
    for (const char* p = linea; *p; p++)
        if (*p == ',') count++;
    return count;
}

double** leer_csv(const char* filename, int* N_out) {
    FILE* fp = fopen(filename, "r");
    if (!fp) {
        perror("Error abriendo el archivo");
        return NULL;
    }

    char* line = malloc(32768);
    int filas = 0, N = 0;

    // Leer primera línea
    if (!fgets(line, 32768, fp)) {
        fprintf(stderr, "Archivo vacío\n");
        fclose(fp);
        free(line);
        return NULL;
    }

    N = contar_columnas(line);
    double** A = allocate_matrix(N);

    // Procesar primera fila
    int col = 0;
    char* token = strtok(line, ",\n");
    while (token && col < N)
        A[0][col++] = atof(token), token = strtok(NULL, ",\n");

    if (col != N) {
        fprintf(stderr, "Error: columnas incorrectas en fila 0\n");
        fclose(fp);
        free_matrix(A, N);
        free(line);
        return NULL;
    }

    filas = 1;

    // Leer el resto
    while (fgets(line, 32768, fp) && filas < N) {
        col = 0;
        token = strtok(line, ",\n");
        while (token && col < N)
            A[filas][col++] = atof(token), token = strtok(NULL, ",\n");
        if (col != N) {
            fprintf(stderr, "Error: columnas incorrectas en fila %d\n", filas);
            fclose(fp);
            free_matrix(A, N);
            free(line);
            return NULL;
        }
        filas++;
    }

    fclose(fp);
    free(line);

    if (filas != N) {
        fprintf(stderr, "Error: filas insuficientes\n");
        free_matrix(A, N);
        return NULL;
    }

    *N_out = N;
    return A;
}

int main() {
    int N;
    double** A = leer_csv("matriz.csv", &N);
    if (!A) return 1;

    double** A_inv = allocate_matrix(N);

    struct timeval start, end;
    gettimeofday(&start, NULL);

    invert_matrix(A, A_inv, N);

    gettimeofday(&end, NULL);
    double t = (end.tv_sec - start.tv_sec) + (end.tv_usec - start.tv_usec) / 1e6;

    printf("Matriz inversa A^-1:\n");
    print_matrix(A_inv, N);
    printf("\nTiempo de ejecución: %.6f segundos\n", t);

    free_matrix(A, N);
    free_matrix(A_inv, N);
    return 0;
}

